controller/index.js

```diff
@@ -1,1019 +1,1054 @@
+    var options = req.query
+    var id = params.id
+    var item = params.item
+    var key = params.key
+
+    agol.find(id, function (err, data) {
+      if (err) {
+        return callback(err, null)
+      }
+
+      // Get the item
+      if (!parseInt(options.layer, 0)) {
+        options.layer = 0
+      }
+
+      agol.getItemData(data.host, id, item, key, options, function (error, itemJson) {
+        if (itemJson && itemJson.koop_status === 'processing' && typeof req.params.silent === 'undefined') {
+          // callback is never called?
+          return controller._returnProcessing(req, res, itemJson, callback)
         }
-      });  
-    };
+        callback(error, itemJson)
+      })
+    })
+  }
+
+  /**
+   * Creates a unique based on request params and the querystring
+   * @returns {string} key
+   */
+  controller._createCacheKey = function (params, query) {
+    // sort the req.query before we hash so we are consistent
+    var sorted_query = {}
+    _(query).keys().sort().each(function (key) {
+      if (key !== 'url_only' && key !== 'format') {
+        sorted_query[key] = query[key]
+      }
+    })
+    // build the file key as an MD5 hash that's a join on the paams and look for the file
+    var toHash = params.item + '_' + (params.layer || 0) + JSON.stringify(sorted_query)
+
+    return crypto.createHash('md5').update(toHash).digest('hex')
+  }
+
+  /**
+   * Finds the item's data
+   * this method is crux of the controller. It handles the logic for returning data in several forms:
+   *   - status processing return 202
+   *   - status is processing but a file exists
+   *   - a file exists for the data
+   *   - a new file needs to be created
+   *   - no format given, just returns json
+   * @param {object} req - the incoming request object
+   * @param {object} res - the outgoing response object
+   */
+  controller.findItemData = function (req, res) {
+    var tableKey = controller._createTableKey('agol', req.params)
+    var dir = req.params.item + '_' + (req.params.layer || 0)
+    var path
 
-    // CHECK the time since our last cache entry 
-    // if > 24 hours since; clear cache and wipe files 
-    // else move on
-    var table_key = ['agol', req.params.item, (req.params.layer || 0)].join(':');
-    agol.getInfo(table_key, function(err, info){
+    // returns data in the data
+    agol.getInfo(tableKey, function (err, info) {
+      if (err) {
+        agol.log('error', err)
+      }
 
-      // parse the spatial ref if we have one, 
+      // parse the spatial ref if we have one,
       // if its whitelisted remove it from the query object
       if (req.query.outSR) {
-        var sr = agol.parseSpatialReference(req.query.outSR);
-        if (sr && sr.wkid && [3785, 3857, 4326, 102100].indexOf(sr.wkid) !== -1){
-          delete req.query.outSR;
+        var sr = agol.parseSpatialReference(req.query.outSR)
+        if (sr && sr.wkid && [3785, 3857, 4326, 102100].indexOf(sr.wkid) !== -1) {
```

Comment:
I think the best place for this to happen is in the ArcGIS Open Data's implementation of Koop rather than in core or in this provider